[FLINK-15794][Kubernetes] Generate the Kubernetes default image version (#11245)
apache/flink: docs/ops/deployment/kubernetes.md

### Session cluster resource definitions

> The Deployment definitions use the pre-built image `flink:{{ site.version }}-scala_2.11` (or `flink:{{ site.version }}-scala_2.12`) which can be found [on Docker Hub](https://hub.docker.com/r/_/flink/).

Reviewer: I am afraid we could not use `{{ site.version }}` here. We have at lease two reasons.
* When the minor version increase, the `{{ site.version }}` do not change. For example, 1.9.2 has released, and the `{{ site.version }}` is still 1.9.0.
* We do not have a image version for the master branch document.

Author: - Strange. I see here that the site.version for 1.9.2 was actually set to 1.9.2 : https://github.com/apache/flink/blob/release-1.9.2/docs/_config.yml#L30 . I do agree that for example https://ci.apache.org/projects/flink/flink-docs-release-1.9/dev/connectors/pubsub.html still shows 1.9.0. To me this seems more like the site needs to be regenerated than this variable having the incorrect value.

- Perhaps on releasing a 'SNAPSHOT' version of the software a 'SNAPSHOT' version docker image with the same software should be released too?

On the other hand ... it is 'just documentation'.
I do see that apparently the 2.12 is 'preferred' so I'm swapping them in the documentation.

Reviewer: Releasing a `SNAPSHOT` image tag is still in discussion. Currently, Flink only provide the image for released versions.

Reviewer 2: As I mentioned this change is not strictly related to the JIRA at hand, as it is about standalone kubernetes deployment and not about the native kubernetes image option.
It seems this `site.version` is not reliable at the moment. I agree with @wangyang0918,
this can bring a confusion.
I still think we should create another issue for this doc change which depends on 2 problems to be resolved: 
- the image Flink version should point to the latest minor version,
- there are no Flink snapshot versions in docker hub at the moment so master doc will point to a not existing image.

Reviewer 2: We had the following idea with @zentol (you can check it in #12131):
```
flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest{% endif %}
```

This way we keep it a concrete version for the stable released docs, like e.g. `1.10.1` (the bug fix release guide should include incrementing minor version in `site.version`, this should change after `1.11`). There is always a docker image for the released versions.

The master docs (not released, like 1.11-SNAPHOST) will always point to the `latest` released docker image for consistency for now until we introduce snapshots for docker image as well.

I know that the implicit scala version is currently `2.12` for short docker tags but it is `2.11` for everything else. `site.scala_version` is also `2.11`. It should have probably been `2.11` for docker tags as well.

> The image is built from this [Github repository](https://github.com/docker-flink/docker-flink).

Author: Yes, I'll update that.

### Session cluster resource definitions

> The Deployment definitions use the pre-built image `flink:{{ site.version }}-scala_2.12` (or `flink:{{ site.version }}-scala_2.11`) which can be found [on Docker Hub](https://hub.docker.com/_/flink/).

Author: As a rule for myself I always try to make the documentation match the change I'm writing.
In this case for me this is really about avoiding the extremely obscure errors you get if you use the wrong Scala version docker image. So in this case to me this documentation seemed like it needed to be fixed as well as part of this change set.

Side note:
Given that the flink docker image must match both the Flink version and Scala version used by my application to run successfully: I would vote to completely drop all docker tags that do not explicitly specify both Flink and Scala versions (i.e. no more `latest` , `scala_2.12` or `1.10.0` tags).

> The Deployment definitions use the pre-built image `flink:{{ site.version }}-scala_2.12` (or `flink:{{ site.version }}-scala_2.11`) which can be found [on Docker Hub](https://hub.docker.com/_/flink/).

Reviewer: I agree with @azagrebin that we should move out this changes. First, it is not related to this ticket(default image version of native K8s). Second, we could not guarantee that it is always existing in the docker hub.

## Appendix

### Common cluster resource definitions
The Deployment definitions use the pre-built image `flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}`{% else %}latest` (which contains the latest released Flink built against Scala 2.12){% endif %} which can be found [on Docker Hub](https://hub.docker.com/_/flink/).

> The image is built from this [Github repository](https://github.com/apache/flink-docker).

Reviewer 2: We already have a chapter on this page about the image, I would rather refer to it: `#flink-docker-image`.
This way we would have to change the image description only in one place. Also, the information about scala 2.12 for latest can become stale, it would make sense to just refer to the tags page. We have everywhere Scala 2.11 by default except docker and might want to make it 2.11 for docker as well later.

Author: Yes, I fixed this. I've added a bit in the central docker documentation files instead.

## Appendix

### Common cluster resource definitions

> 

Author: Yes

    spec:
      containers:
      - name: jobmanager

>         image: flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest # This contains the latest released Flink built against Scala 2.12{% endif %}

Author: Yes, I'll reword to make it better maintainable.

>         image: flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest # This contains the latest released Flink built against Scala 2.12{% endif %}

Author: I came up with the generic rewording (which only is shown if the documentation is built for a snapshot version which will show 'latest'):
``` The 'latest' tag contains the latest released version of Flink for a specific Scala version which will mismatch with your application over time.```

    spec:
      containers:
      - name: jobmanager

>         image: flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest # The 'latest' tag contains the latest released version of Flink for a specific Scala version which will mismatch with your application over time.{% endif %}

Reviewer 2: ```suggestion
        image: flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest # The 'latest' tag contains the latest released version of Flink for a specific Scala version which can mismatch with the Scala version of your application.{% endif %}
```
Will the Scala version definitely mismatch?
I am wondering why this comment should be in all places. This is for dev docs.
Could we not put this comment to that note in the relevant docker/tags chapter?

Author: In my experience: Yes, the versions will mismatch at some point. 
Also with a Scala mismatch the problem is big, yet with a Flink version mismatch I suspect more subtle problems to occur.

Also I consider the choice of having Scala 2.12 the version used for "latest" a choice for today which may change (As I mentioned in the mailing list I am actually against having a "latest" for something like Flink).

Note that this comment is ONLY shown when the documentation is generated for a non-stable "SNAPSHOT" build of the documentation. So the normal documentation should not show this.
I duplicated it in all of those places because these are Kubernetes config files I expect to be used by operations people who will have limited knowledge of the application at hand.

Reviewer 2: Alright but if we and user always use the same Scala version, it will never mismatch, right?

Author: For me one big trigger was that I had a mis match and the errors were not helping me figure out what was happening.

Reviewer 2: This sounds fine for me to fix.
It just sounds now as if it `will` definitely mismatch, although it looks like it `can` mismatch if not aligned with the image.

Reviewer 2: How about this:
```
flink:latest # The 'latest' tag contains the latest released version of Flink for a specific Scala version which can conflict with the versions, used by your application.
```
Somehow this sounds confusing:
`which will mismatch with your application over time`, in particular `over time`.

Reviewer 3: I think we should be way more explicit about this; scala users should _never_ use `latest`; it is simply to easy to run into issues. If we release a newer release, with scala 2.13 support we suddenly break everything.
I would explicitly phrase it like that; use it only if you are only using java, otherwise select a specific Flink version with a matching scala version.

We should also look into providing a `latest_<scala.version>` tag to make this easier for scala users.

Reviewer 2: We added a [note](https://github.com/apache/flink/pull/11245/files#diff-a530b60989203f9e1f03c64c57deb56cR49) for this which can be promoted to warning.
I think the idea here is to keep the comment in this particular line (and others like this) brief enough as it is a code example.
